lib/slug.rb

```diff
@@ -1,21 +1,35 @@
 # encoding: utf-8
 
-
 module Slug
 
-  def self.for(string)
-    # TODO review if ja should use this
-    # ko asked for it to be removed
-    if ['zh_CN', 'ja'].include?(SiteSetting.default_locale)
-      unless defined? Stringex
-        require 'stringex_lite'
-      end
-      slug = string.to_url
-    else
-      slug = string.gsub("'", "").parameterize
-      slug.gsub!("_", "-")
-    end
-    slug =~ /[^\d]/ ? slug : '' # Reject slugs that only contain numbers, because they would be indistinguishable from id's.
+  def self.for(string, default = 'topic')
+    slug = case SiteSetting.slug_generation_method.to_sym
+           when :ascii then self.ascii_generator(string)
+           when :encoded then self.encoded_generator(string)
+           when :none then self.none_generator(string)
+           else raise Discourse::SiteSettingMissing
+           end
+    # Reject slugs that only contain numbers, because they would be indistinguishable from id's.
+    slug = (slug =~ /[^\d]/ ? slug : '')
+    slug.blank? ? default : slug
+  end
+
+  private
+
+  def self.ascii_generator(string)
+    slug = string.gsub("'", "").parameterize
+    slug.gsub("_", "-")
   end
 
+  def self.encoded_generator(string)
+    # This regexp sanitized special characters, including reserved characters from RFC3986.
+    # :/?#[]@!$&'()*+,;=_.~%\`^|{}"<>
+    # See also URI::REGEXP::PATTERN.
+    string = string.strip.gsub(/\s+/, '-').gsub(/[:\/\?#\[\]@!\$&'\(\)\*\+,;=_\.~%\\`^\s|\{\}"<>]+/, '')
```

Comment:
@ZogStriP URI::REGEXP::PATTERN follows RFC 2396. This regex is a super set of it. Almost all special characters are sanitized by this regex. It's a convenient way to avoid any edge cases.